radar/src/radar.rs

```diff
@@ -0,0 +1,603 @@
+        frame.entity_id,
+        super::POSITION
+    ))?;
+
+    let resource_id = format!("decs.components.{}.{}", frame.shard, frame.entity_id);
+
+    if let (Some(radar_str), Some(position_str)) = (radar_receiver_value, position_value) {
+        let radar_receiver: RadarReceiver = serde_json::from_str(&radar_str)?;
+        let position: Position = serde_json::from_str(&position_str)?;
+
+        let radar_contacts_key = &format!(
+            "decs:components:{}:{}:{}",
+            frame.shard, frame.entity_id, RADAR_CONTACTS
+        );
+
+        let old_contacts: HashMap<String, RadarContact> = if ctx.kv().exists(radar_contacts_key)? {
+            let contact_members = ctx.kv().set_members(radar_contacts_key)?;
+            let mut old_contacts: HashMap<String, RadarContact> = HashMap::new();
+            for c in contact_members {
+                let contact = c.replace(".", ":");
+                if let Some(radar_contact_str) = ctx.kv().get(&contact)? {
+                    old_contacts.insert(
+                        contact,
+                        serde_json::from_str(&radar_contact_str.to_string())?,
+                    );
+                }
+            }
+            old_contacts
+        } else {
+            HashMap::new()
+        };
+
+        let all_positions = POSITIONS.read().unwrap();
+        let updates = radar_updates(
+            &frame.entity_id,
+            &position,
+            &radar_receiver,
+            &old_contacts,
+            &all_positions,
+        );
+
+        for update in updates {
+            match update {
+                RadarContactDelta::Add(rc) => {
+                    let new_subject = ResProtocolRequest::New(format!(
+                        "{}.{}",
+                        resource_id.to_string(),
+                        RADAR_CONTACTS
+                    ))
+                    .to_string();
+                    let payload = serde_json::json!({ "params": rc });
+                    publish_message(ctx, &new_subject, payload)?;
+                }
+                RadarContactDelta::Remove(rid) => {
+                    let delete_subject = ResProtocolRequest::Delete(format!(
+                        "{}.{}",
+                        resource_id.to_string(),
+                        RADAR_CONTACTS
+                    ))
+                    .to_string();
+                    let payload = serde_json::json!({ "params": {"rid": rid.replace(":", ".")}});
+                    publish_message(ctx, &delete_subject, payload)?;
+                }
+                RadarContactDelta::Change(rid, rc) => {
+                    let set_subject = &format!("call.{}.set", rid);
+                    let payload = serde_json::json!({ "params": rc });
+                    publish_message(ctx, set_subject, payload)?;
+                }
+            }
+        }
+    }
+    Ok(vec![])
+}
+
+/// Helper function used to publish a payload on a specified subjct
+fn publish_message(
```

Comment:
I feel like this function doesn't belong here, is more of a general function that other systems could use too.

Reply:
right now it's just a few lines of convenience code. Let's avoid "prefactoring" this into some kind of shared utility until we have a better idea of the kinds of SDK-like utilities system builders will need.